compiler/qsc_data_structures/src/language_features.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use clap::ValueEnum;
+use serde::Deserialize;
+use std::collections::BTreeSet;
+#[derive(Deserialize, Clone, Debug, PartialOrd, Ord, PartialEq, Eq, Copy)]
+#[serde(rename_all = "kebab-case")]
+pub enum LanguageFeature {
+    /// This language feature enables experimental syntax that will likely be stabilized in the next major version.
+    /// It may include removals of outdated syntax and introductions of new syntax.
+    V2PreviewSyntax,
+}
+
+impl LanguageFeature {
+    pub fn try_parse(s: &str) -> Result<Self, UnrecognizedLanguageFeature> {
+        match s {
+            "v2-preview-syntax" => Ok(LanguageFeature::V2PreviewSyntax),
+            _ => Err(UnrecognizedLanguageFeature::msg(format!(
+                "Unrecognized language feature: {s}"
+            ))),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Default)]
+pub struct LanguageFeatures(BTreeSet<LanguageFeature>);
+pub type LanguageFeatureIncompatibility = miette::ErrReport;
+pub type UnrecognizedLanguageFeature = miette::ErrReport;
+
+impl LanguageFeatures {
+    /// Checks that the current set of language features is compatible and well-formed.
+    /// Returns a descriptive error message if not.
+    pub fn check_compatibility(&self) -> Result<(), LanguageFeatureIncompatibility> {
+        // we currently only have one language feature, so no checking is required.
+        Ok(())
+    }
+
+    #[must_use]
+    pub fn none() -> Self {
```

Comment:
I think it would be better to have a `default()` method instead of a `none()` method. Since we can use language features to experiment with things that might not be ready for release but are not breaking changes (e.g. performing RCA), we could release these features by including them in `default()`.